src/ops.rs

```diff
@@ -707,21 +746,338 @@ pub fn imul3_32bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -
+    }
+    vm.set_arg(pipeline.args[0].location, SizedValue::Dword(result as u32))?;
+    Ok(())
+}
+
+#[allow(exceeding_bitshifts)]
+pub fn ror_8bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    let destination = vm.get_arg(pipeline.args[0].location)?.u8_exact()?;
+    let mut count = vm.get_arg(pipeline.args[1].location)?.u8_exact()?;
+    count = count & 0x1F;
+    count = count % 8;
+    if count == 0 {
+        return Ok(());
+    }
+    let result = (destination >> count) | (destination << (8 - count));
+    vm.flags.carry = result.get::<BigEndian>(0.into()) as bool;
+    if count == 1 {
+        vm.flags.overflow = vm.flags.carry ^ ((result & 0x40) > 0);
+    }
+    vm.set_arg(pipeline.args[0].location, SizedValue::Byte(result))?;
+    Ok(())
+}
+
+pub fn ror_native_word(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    if pipeline.size_override {
+        return ror_16bit(vm, pipeline, _hv);
+    } else {
+        return ror_32bit(vm, pipeline, _hv);
+    }
+}
+
+pub fn ror_16bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    let destination = vm.get_arg(pipeline.args[0].location)?.u16_exact()?;
+    let mut count = vm.get_arg(pipeline.args[1].location)?.u16_sx()?;
+    count = count & 0x1F;
+    count = count % 16;
+    if count == 0 {
+        return Ok(());
+    }
+    let result = (destination >> count) | (destination << (16 - count));
+    vm.flags.carry = result.get::<BigEndian>(0.into()) as bool;
+    if count == 1 {
+        vm.flags.overflow = vm.flags.carry ^ ((result & 0x4000) > 0);
+    }
+    vm.set_arg(pipeline.args[0].location, SizedValue::Word(result))?;
+    Ok(())
+}
+
+pub fn ror_32bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    let destination = vm.get_arg(pipeline.args[0].location)?.u32_exact()?;
+    let mut count = vm.get_arg(pipeline.args[1].location)?.u32_sx()?;
+    count = count & 0x1F;
+    if count == 0 {
+        return Ok(());
+    }
+    let result = (destination >> count) | (destination << (32 - count));
+    vm.flags.carry = result.get::<BigEndian>(0.into()) as bool;
+    if count == 1 {
+        vm.flags.overflow = vm.flags.carry ^ ((result & 0x40000000) > 0);
+    }
+    vm.set_arg(pipeline.args[0].location, SizedValue::Dword(result))?;
+    Ok(())
+}
+
+pub fn sar_8bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    let mut destination = vm.get_arg(pipeline.args[0].location)?.u8_exact()?;
+    let mut count = vm.get_arg(pipeline.args[1].location)?.u8_exact()?;
+    if count == 1 {
+        vm.flags.overflow = false;
+    }
+    count &= 0x1F;
+    if count == 0 {
+        return Ok(());
+    }
+    vm.flags.carry = ((destination >> (count - 1)) & 1) > 0;
+    // get MSB
+    if destination.get::<BigEndian>(0.into()) as bool {
```

Comment:
this can be simplified to `if destination & 0x80 > 0` though I don't have a strong opinion on which way is more clear. The constant switching between BigEndian and LittleEndian throughout this file for picking out bits is confusing though

Reply:
yea there was some confusion with that with these opcodes but I figured it out. I can try to make them a bit more consistent.